src/core/ddsi/src/ddsi_init.c

```diff
@@ -1587,23 +1683,25 @@ int ddsi_init (struct ddsi_domaingv *gv, struct ddsi_psmx_instance_locators *psm
 #endif
 
   gv->as_disc = ddsi_new_addrset ();
-  if (gv->config.allowMulticast & DDSI_AMC_SPDP)
-    ddsi_add_locator_to_addrset (gv, gv->as_disc, &gv->loc_spdp_mc);
-  /* If multicast was enabled but not available, always add the local interface to the discovery address set.
-     Conversion via string and add_peer_addresses has the benefit that the port number expansion happens
-     automatically. */
   for (int i = 0; i < gv->n_interfaces; i++)
   {
-    if (!gv->interfaces[i].mc_capable && gv->config.peers == NULL)
+    if ((gv->interfaces[i].allow_multicast & DDSI_AMC_SPDP) &&
+        ddsi_factory_supports (gv->xmit_conns[i]->m_factory, gv->loc_spdp_mc.kind))
     {
-      struct ddsi_config_peer_listelem peer_local;
-      char local_addr[DDSI_LOCSTRLEN];
-      ddsi_locator_to_string_no_port (local_addr, sizeof (local_addr), &gv->interfaces[i].loc);
-      peer_local.next = NULL;
-      peer_local.peer = local_addr;
-      add_peer_addresses (gv, gv->as_disc, &peer_local);
+      const ddsi_xlocator_t xloc = { .conn = gv->xmit_conns[i], .c = gv->loc_spdp_mc };
+      ddsi_add_xlocator_to_addrset (gv, gv->as_disc, &xloc);
     }
   }
+  if (add_self_to_as_disc)
+  {
+    struct ddsi_config_peer_listelem peer_local;
+    char local_addr[DDSI_LOCSTRLEN];
+    ddsi_locator_to_string_no_port (local_addr, sizeof (local_addr), &gv->interfaces[0].loc);
```

Comment:
I think the changes in 65a8843 take care of the first part.

The second question: yes, I believe `interfaces[0]` is ok. There is currently always at least one real network interface and by construction that one is in `interfaces[0]`. It think it is a reasonable assumption that it can always be used to send data to itself. If that's not true I think we might be looking at essentially another version of the (new) check to see if loopback multicast works on linux.